File: app/src/main/java/com/example/finalexer3grp2/EditFilmFragment.java
package com.example.finalexer3grp2;

import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import com.example.finalexer3grp2.databinding.FragmentEditFilmBinding;
import com.google.android.material.chip.Chip;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import coil.Coil;
import coil.request.ImageRequest;

public class EditFilmFragment extends Fragment {

    private FragmentEditFilmBinding binding;
    private int filmId = -1;
    private Film currentFilm;
    private Uri selectedImageUri;

    private final ActivityResultLauncher<String> pickImage = registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    selectedImageUri = uri;
                    binding.ivEditPoster.setImageURI(uri);
                }
            }
    );

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentEditFilmBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (getArguments() != null) {
            filmId = getArguments().getInt("filmId", -1);
        }

        if (filmId != -1) loadFilm();

        binding.toolbarEditFilm.setNavigationOnClickListener(v -> Navigation.findNavController(view).popBackStack());

        binding.cvEditPoster.setOnClickListener(v -> pickImage.launch("image/*"));

        binding.btnUpdateFilm.setOnClickListener(v -> updateFilm());
    }

    private void loadFilm() {
        FilmDatabase.databaseWriteExecutor.execute(() -> {
            FilmDatabase db = FilmDatabase.getDatabase(requireContext());
            currentFilm = db.filmDao().getFilmById(filmId);

            if (currentFilm != null && getActivity() != null) {
                getActivity().runOnUiThread(() -> {
                    binding.etEditTitle.setText(currentFilm.getTitle());
                    binding.etEditYear.setText(currentFilm.getYear());
                    binding.etEditDirector.setText(currentFilm.getDirector());
                    binding.etEditDescription.setText(currentFilm.getDescription());

                    if (currentFilm.getPosterUri() != null) {
                        selectedImageUri = Uri.parse(currentFilm.getPosterUri());
                        ImageRequest request = new ImageRequest.Builder(requireContext())
                                .data(selectedImageUri)
                                .target(binding.ivEditPoster)
                                .build();
                        Coil.imageLoader(requireContext()).enqueue(request);
                    }

                    List<String> genres = Arrays.asList(currentFilm.getGenres().split(", "));
                    for (int i = 0; i < binding.cgEditGenres.getChildCount(); i++) {
                        Chip chip = (Chip) binding.cgEditGenres.getChildAt(i);
                        chip.setChecked(genres.contains(chip.getText().toString()));
                    }
                });
            }
        });
    }

    private void updateFilm() {
        String title = binding.etEditTitle.getText().toString();
        String year = binding.etEditYear.getText().toString();

        if (title.trim().isEmpty() || year.trim().isEmpty()) {
            Toast.makeText(requireContext(), "Required fields missing", Toast.LENGTH_SHORT).show();
            return;
        }

        List<String> selectedGenres = new ArrayList<>();
        for (int i = 0; i < binding.cgEditGenres.getChildCount(); i++) {
            Chip chip = (Chip) binding.cgEditGenres.getChildAt(i);
            if (chip.isChecked()) selectedGenres.add(chip.getText().toString());
        }

        if (currentFilm != null) {
            currentFilm.setTitle(title);
            currentFilm.setYear(year);
            currentFilm.setDirector(binding.etEditDirector.getText().toString());
            currentFilm.setDescription(binding.etEditDescription.getText().toString());
            currentFilm.setGenres(selectedGenres.isEmpty() ? "Uncategorized" : String.join(", ", selectedGenres));
            currentFilm.setPosterUri(selectedImageUri != null ? selectedImageUri.toString() : null);

            currentFilm.setUpdatedAt(System.currentTimeMillis());

            FilmDatabase.databaseWriteExecutor.execute(() -> {
                FilmDatabase.getDatabase(requireContext()).filmDao().updateFilm(currentFilm);
                if (getActivity() != null) {
                    getActivity().runOnUiThread(() -> {
                        Toast.makeText(requireContext(), "Updated!", Toast.LENGTH_SHORT).show();
                        Navigation.findNavController(requireView()).popBackStack();
                    });
                }
            });
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/finalexer3grp2/EditFilmFragment.kt
package com.example.finalexer3grp2

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import coil.load
import com.example.finalexer3grp2.databinding.FragmentEditFilmBinding
import com.google.android.material.chip.Chip
import kotlinx.coroutines.launch

class EditFilmFragment : Fragment() {

    private var _binding: FragmentEditFilmBinding? = null
    private val binding get() = _binding!!

    private var filmId = -1
    private var currentFilm: Film? = null
    private var selectedImageUri: Uri? = null

    private val pickImage =
        registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
            uri?.let {
                selectedImageUri = it
                binding.ivEditPoster.setImageURI(it)
            }
        }

    override fun onCreateView(
        inflater: android.view.LayoutInflater,
        container: android.view.ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentEditFilmBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        filmId = arguments?.getInt("filmId") ?: -1

        if (filmId != -1) loadFilm()

        binding.toolbarEditFilm.setNavigationOnClickListener {
            findNavController().popBackStack()
        }

        binding.cvEditPoster.setOnClickListener {
            pickImage.launch("image/*")
        }

        binding.btnUpdateFilm.setOnClickListener {
            updateFilm()
        }
    }

    private fun loadFilm() {
        lifecycleScope.launch {
            val db = FilmDatabase.getDatabase(requireContext())
            currentFilm = db.filmDao().getFilmById(filmId)

            currentFilm?.let { film ->
                binding.etEditTitle.setText(film.title)
                binding.etEditYear.setText(film.year)
                binding.etEditDirector.setText(film.director)
                binding.etEditDescription.setText(film.description)

                selectedImageUri = film.posterUri?.let { Uri.parse(it) }
                binding.ivEditPoster.load(selectedImageUri)

                val genres = film.genres.split(", ")
                for (i in 0 until binding.cgEditGenres.childCount) {
                    val chip = binding.cgEditGenres.getChildAt(i) as Chip
                    chip.isChecked = genres.contains(chip.text.toString())
                }
            }
        }
    }

    private fun updateFilm() {
        val title = binding.etEditTitle.text.toString()
        val year = binding.etEditYear.text.toString()

        if (title.isBlank() || year.isBlank()) {
            Toast.makeText(requireContext(), "Required fields missing", Toast.LENGTH_SHORT).show()
            return
        }

        val selectedGenres = mutableListOf<String>()
        for (i in 0 until binding.cgEditGenres.childCount) {
            val chip = binding.cgEditGenres.getChildAt(i) as Chip
            if (chip.isChecked) selectedGenres.add(chip.text.toString())
        }

        val updated = currentFilm?.copy(
            title = title,
            year = year,
            director = binding.etEditDirector.text.toString(),
            description = binding.etEditDescription.text.toString(),
            genres = if (selectedGenres.isEmpty()) "Uncategorized"
            else selectedGenres.joinToString(", "),
            posterUri = selectedImageUri?.toString()
        )

        lifecycleScope.launch {
            updated?.let {
                FilmDatabase.getDatabase(requireContext())
                    .filmDao()
                    .updateFilm(it)

                Toast.makeText(requireContext(), "Updated!", Toast.LENGTH_SHORT).show()
                findNavController().popBackStack()
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}